File: native/ios/Sources/AppState.swift
import SwiftUI

/// Holds the enrolled identity + token, persisted to UserDefaults so the app
/// resumes signed-in. Drives the root switch between Welcome and the tab bar.
@MainActor
final class AppState: ObservableObject {
    @Published var uid: String?
    @Published var token: String?
    @Published var displayName: String = ""
    // The user's chosen language also drives the app chrome via L10n.
    @Published var language = "en"


    /// The person's own model key, held on this device only and sent per
    /// request as `x-llm-api-key`.
    ///
    /// The console has offered this since 0.4.3 and the phones never did — so
    /// somebody who set their key in the console had it used there and the
    /// deployment's key used here, on the same profile, with nothing saying
    /// so. Never in the account and never on the wire except as the header.
    @Published var llmKey = ""
    @Published var signupKey = ""

    private let d = UserDefaults.standard

    init() {
        uid = d.string(forKey: "jim.uid")
        token = d.string(forKey: "jim.token")
        displayName = d.string(forKey: "jim.name") ?? ""
        language = d.string(forKey: "jim.lang") ?? "en"
        llmKey = d.string(forKey: "jim.llmKey") ?? ""
        let held = llmKey
        Task { await ApiClient.shared.useLlmKey(held) }
        signupKey = d.string(forKey: "jim.signupKey") ?? ""
        let invite = signupKey
        Task { await ApiClient.shared.useSignupKey(invite) }
    }

    var isEnrolled: Bool { uid != nil && token != nil }

    func signIn(_ r: EnrollResult) {
        uid = r.id; token = r.user_token; displayName = r.display_name
        d.set(r.id, forKey: "jim.uid")
        d.set(r.user_token, forKey: "jim.token")
        d.set(r.display_name, forKey: "jim.name")
    }

    func signOut() {
        uid = nil; token = nil; displayName = ""
        ["jim.uid", "jim.token", "jim.name", "jim.lang"].forEach { d.removeObject(forKey: $0) }
    }

    /// Store or clear it. An empty string is the clear: there is no flag to
    /// leave switched on by mistake, and no key means the deployment's.
    func rememberLlmKey(_ key: String) {
        let trimmed = key.trimmingCharacters(in: .whitespacesAndNewlines)
        llmKey = trimmed
        if trimmed.isEmpty { d.removeObject(forKey: "jim.llmKey") }
        else { d.set(trimmed, forKey: "jim.llmKey") }
        Task { await ApiClient.shared.useLlmKey(trimmed) }
    }

    /// The deployment invite key, same clearing rule: empty means none, and
    /// a deployment that never gated signup never needs one.
    func rememberSignupKey(_ key: String) {
        let trimmed = key.trimmingCharacters(in: .whitespacesAndNewlines)
        signupKey = trimmed
        if trimmed.isEmpty { d.removeObject(forKey: "jim.signupKey") }
        else { d.set(trimmed, forKey: "jim.signupKey") }
        Task { await ApiClient.shared.useSignupKey(trimmed) }
    }

    func rememberLanguage(_ code: String) {
        language = code
        d.set(code, forKey: "jim.lang")
    }
}

File: native/windows/AppState.cs
using System;
using System.IO;
using System.Text.Json;

namespace JimGuardian;

/// <summary>
/// The enrolled identity + token, persisted to a small JSON file under
/// LocalApplicationData so the app resumes signed-in (unpackaged-safe).
/// </summary>
public sealed class AppState
{
    public static AppState Current { get; } = Load();

    public string? Uid { get; set; }
    public string? Token { get; set; }
    public string DisplayName { get; set; } = "";
    // The user's chosen language also drives the app chrome via L10n.
    public string Language { get; set; } = "en";

    /// <summary>The person's own model key, held on this machine only and
    /// sent per request as <c>x-llm-api-key</c>.
    ///
    /// <para>The console has offered this since 0.4.3 and the shells never
    /// did — so a key set in the console was used there and the deployment's
    /// key used here, on the same profile, with nothing saying so.</para>
    /// </summary>
    public string LlmKey { get; set; } = "";

    /// <summary>Store or clear it. Empty is the clear: no key means the
    /// deployment's, and there is no flag to leave switched on by
    /// mistake.</summary>
    public void RememberLlmKey(string key)
    {
        LlmKey = (key ?? "").Trim();
        Save();
    }

    /// <summary>The deployment invite key, sent per request as
    /// <c>x-signup-key</c>. A published deployment gates account creation
    /// behind one; empty means none, which is what every ungated
    /// deployment wants.</summary>
    public string SignupKey { get; set; } = "";

    public void RememberSignupKey(string key)
    {
        SignupKey = (key ?? "").Trim();
        Save();
    }

    public void RememberLanguage(string code)
    {
        Language = code;
        Save();
    }

    public bool IsEnrolled => !string.IsNullOrEmpty(Uid) && !string.IsNullOrEmpty(Token);

    private static string PathOnDisk =>
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                     "JimGuardian", "session.json");

    public void SignIn(EnrollResult r)
    {
        Uid = r.Id; Token = r.UserToken; DisplayName = r.DisplayName;
        Save();
    }

    public void SignOut()
    {
        Uid = null; Token = null; DisplayName = "";
        try { File.Delete(PathOnDisk); } catch { /* ignore */ }
    }

    private void Save()
    {
        Directory.CreateDirectory(Path.GetDirectoryName(PathOnDisk)!);
        File.WriteAllText(PathOnDisk, JsonSerializer.Serialize(this));
    }

    private static AppState Load()
    {
        try
        {
            if (File.Exists(PathOnDisk))
                return JsonSerializer.Deserialize<AppState>(File.ReadAllText(PathOnDisk)) ?? new AppState();
        }
        catch { /* fall through to fresh state */ }
        return new AppState();
    }
}
